### PR TITLE
NOXSolver: add a test demonstrating the issue with pending_exception

### PR DESCRIPTION
**Pull request by vovannikov**
The bug, which we encountered in our code and which @peterrum mentioned in #16028, is in fact not related to function `clear()` but somewhat related to the the recoverability of NOX.

The callback for `solve_with_jacobian` does not _necessarily_ interrupts the current NOX iteration. NOX does so only if the setting `Rescue Bad Newton Solve` is set to `false`. Otherwise it will proceed to the next Newton step even if an exception was thrown `solve_with_jacobian` (for instance, if the number of GMRES iterations is exceeded) and `-1` was returned by `call_and_possibly_capture_exception()`. That is exactly what happens in our code. The worst part is that by default `Rescue Bad Newton Solve = true`.

This PR just shows the issue but does not solve it. As a possible solution, I propose to remove this assert here: https://github.com/dealii/dealii/blob/29f9da0a3412c3530b2f1b93c0f59ab6d909546f/include/deal.II/trilinos/nox.templates.h#L66

### CODE REVIEW

File: tests/trilinos/nox_solver_05.cc [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  non_linear_parameters->set("Nonlinear Solver", "Line Search Based");
+  non_linear_parameters->sublist("Printing").set("Output Information", 15);
+  non_linear_parameters->sublist("Direction").set("Method", "Newton");
+  non_linear_parameters->sublist("Direction")
+    .sublist("Newton")
+    .sublist("Linear Solver")
+    .set("Tolerance", lin_rel_tolerance);
+  non_linear_parameters->sublist("Direction")
+    .sublist("Newton")
+    .set("Rescue Bad Newton Solve", false);
```
Comment (vovannikov) [inline tests/trilinos/nox_solver_05.cc:71]: The test is valid and passes but will get broken if the following lines are commented out.

**Comment by bangerth** [comment]
So you are saying that if the `Rescue Bad Newton Solve` flag is set to:

- `false`, NOX will terminate the iteration if a solver throws an exception
- `true` (the default), NOX will just ignore the exception and try again, perhaps with a smaller step length?

If that is so, then we have two options:

- We just always set that flag to `false`.
- We let NOX do its thing, which is to recover from bad solves where possible, and then have to copy the logic over from SUNDIALS KINSOL wrappers which know how to do this kind of thing.

Does NOX document the ability to recover? Does it allow user callbacks to distinguish between recoverable and non-recoverable errors, for example by using different (nonzero) return values?

**Comment by vovannikov** [comment]
> So you are saying that if the `Rescue Bad Newton Solve` flag is set to:
> 
> * `false`, NOX will terminate the iteration if a solver throws an exception
> * `true` (the default), NOX will just ignore the exception and try again, perhaps with a smaller step length?

Not exactly, it will not event try a smaller step or something like that. The documentation of NOX is quite scarce, but as I understood, in this case NOX simply goes to the next non-linear iteration like everything was fine, only a warning is issued at some verbosity levels. The documentation says (https://docs.trilinos.org/dev/packages/nox/doc/html/classNOX_1_1Direction_1_1Newton.html):

_"Rescue Bad %Newton Solve" (Boolean) - If set to true, we will use the computed direction even if the linear solve does not achieve the tolerance specified by the forcing term. Defaults to true._

That is how it looks in the code:

https://github.com/trilinos/Trilinos/blob/6779ba1c05c50ccb12d78d6b04db737d8e3e4a2f/packages/nox/src/NOX_Direction_Newton.C#L140-L152

> If that is so, then we have two options:
> 
> * We just always set that flag to `false`.

This is an option. 

> Does NOX document the ability to recover? Does it allow user callbacks to distinguish between recoverable and non-recoverable errors, for example by using different (nonzero) return values?

It is a good question. What type of recovery? And at what stage?

The recovery that happens here is not even a recovery, I would say with this setting enable NOX simply ignores the errors from the linear solver and keeps on trying to solve the current non-linear step. However, since different members of `NOX::Abstract::Group`, which the deal.II's `NOXWrappers::Group` is based on, may return different `NOX::Abstract::Group::ReturnType`s, NOX probably may contain some recover mechanism (https://github.com/trilinos/Trilinos/blob/master/packages/nox/src/NOX_Abstract_Group.C). Probably, @peterrum should have more info on that.

**Comment by bangerth** [comment]
"in this case NOX simply goes to the next non-linear iteration like everything was fine, only a warning is issued at some verbosity levels" -- this kind of behavior does not seem to be in the spirit of deal.II. It also does not seem like a good algorithmic choice to "simply go[es] to the next non-linear iteration like everything was fine". I would vote to just unconditionally set that flag to `false` and tell NOX to behave like a reasonable and predictable algorithm.

**Comment by vovannikov** [comment]
> "in this case NOX simply goes to the next non-linear iteration like everything was fine, only a warning is issued at some verbosity levels" -- this kind of behavior does not seem to be in the spirit of deal.II. It also does not seem like a good algorithmic choice to "simply go[es] to the next non-linear iteration like everything was fine". I would vote to just unconditionally set that flag to `false` and tell NOX to behave like a reasonable and predictable algorithm.

I think it is a bit too restrictive to disable this option completely and can be also quite confusing for the user, why the recovery is disabled on the wrapper level. After thinking, I decided to give a try to a bit different solution, please have a look.

The option `Rescue Bad Newton Solve` changes the logic of the solver in qualitative way, so it makes sense that our wrapper should be aware of this information. In this case, I think it is OK to use it and nullify the pending exception if recovery is enabled. With this implementation, we still keep the existing exceptions mechanism allowing user to use the recovery option if needed.

The 2 tests then check the behavior of NOX with the recovery enabled and disabled.

File: include/deal.II/trilinos/nox.templates.h [hunk 2]
```diff
@@ -1089,8 +1092,25 @@ namespace TrilinosWrappers
...
+        // returns code different from 0, then NOX does not interrupt the
+        // solution process but rather performs a recovery step. To ensure this
+        // feature is available to the user, we need to supress the exception in
+        // this case, since it is exactly that, what NOX expects from our
+        // callbacks.
+        const bool do_rescue =
+          parameters->sublist("Newton").get("Rescue Bad Newton Solve", true);
+
+        if (do_rescue && pending_exception != nullptr)
+          pending_exception = nullptr;
```
Comment (bangerth) [inline include/deal.II/trilinos/nox.templates.h:1111]: This isn't right. First, by setting `pending_exception=nullptr`, you are simply abandoning the exception. This is a resource leak. You need to rethrow the exception, catch it, and ignore it if you want to get rid of it. Take a look at how this is done in `kinsol.cc`:
```
    if (pending_exception)
      {
        try
          {
            std::rethrow_exception(pending_exception);
          }
        catch (const RecoverableUserCallbackError &exc)
          {
            pending_exception = nullptr;
            if (status == 0)
              /* just eat the exception */;
            else
              throw;
          }
        catch (...)
          {
            pending_exception = nullptr;
            throw;
          }
      }
```
Second, you only want to do that if the captured exception is of type `RecoverableUserCallbackError`. For every other error, I will still argue that NOX needs to be made to abort its operations and return to user space.

Comment (bangerth) [reply include/deal.II/trilinos/nox.templates.h:1111]: Finally, if we do capture exceptions in these lambda functions, we need to deal with this at the point where we call NOX. Again, kinsol is the right place to look: https://github.com/dealii/dealii/blob/master/source/sundials/kinsol.cc#L475-L505

**Review by bangerth — COMMENTED** [review]
I'm ok with making the behavior depend on certain flags, though I fear that NOX is not flexible enough to actually understand when it can or cannot recover.

File: include/deal.II/trilinos/nox.templates.h (hunk 2 above)
Comment (vovannikov) [reply include/deal.II/trilinos/nox.templates.h:1111]: > This isn't right. First, by setting `pending_exception=nullptr`, you are simply abandoning the exception. This is a resource leak. You need to rethrow the exception, catch it, and ignore it if you want to get rid of it. Take a look at how this is done in `kinsol.cc`:

> Second, you only want to do that if the captured exception is of type `RecoverableUserCallbackError`. For every other error, I will still argue that NOX needs to be made to abort its operations and return to user space.

Both points sound reasonable, I will update the code accordingly.

> Finally, if we do capture exceptions in these lambda functions, we need to deal with this at the point where we call NOX. Again, kinsol is the right place to look: https://github.com/dealii/dealii/blob/master/source/sundials/kinsol.cc#L475-L505

I am not sure I got it right. In these lines indeed there is an exception handling after calling `KINSol()`, but there is a similar thing happening inside `SUNDIALS::Utilities::call_and_possibly_capture_exception()` for every callback call: https://github.com/dealii/dealii/blob/8ad845fa53f571021c5919b6f0842f7a3cb90d2a/include/deal.II/sundials/utilities.h#L57-L73

File: include/deal.II/trilinos/nox.templates.h [hunk 3]
```diff
@@ -1089,8 +1092,40 @@ namespace TrilinosWrappers
...
+              }
+            catch (...)
+              {
+                pending_exception = nullptr;
+                throw;
+              }
           }
+
+        return ret_code;
       }));
```
Comment (bangerth) [inline include/deal.II/trilinos/nox.templates.h:1129]: Yes, this looks good now. Just one minor addition of a comment, and some re-arranging:
```suggestion
        const bool do_rescue =
          parameters->sublist("Newton").get("Rescue Bad Newton Solve", true);

        if (do_rescue && (pending_exception != nullptr))
          {
            try
              {
                std::rethrow_exception(pending_exception);
              }
            catch (const RecoverableUserCallbackError &exc)
              {
                pending_exception = nullptr;
                
                // If the callback threw a recoverable exception, and if recovery is enabled,
                // then eat the exception and return the error code
                return ret_code;
              }
            catch (...)
              {
                // If not a recoverable exception, then just re-throw the exception and hope that
                // NOX knows what to do with propagating exceptions (i.e., does not create a
                // resource leak, for example).
                pending_exception = nullptr;
                throw;
              }
          }
        else
          // Rescue not allowed, or there was no exception -> simply return
          // the value produced by the callback.
          return ret_code;
      }));
```

**Review by bangerth — COMMENTED** [review]
Yes, nice. Just a minor bit of code hygiene. Thanks for sticking with me through these multiple rounds of edits!

Would you mind adding to the documentation of `NOXSolver::solve_with_jacobian` that recovery is possible with the right flag set? Also, you might want to put a changelog entry into `doc/news/changes/minor`! Let me know once you're done with this and then I can merge!

**Comment by vovannikov** [comment]
> Yes, nice. Just a minor bit of code hygiene. Thanks for sticking with me through these multiple rounds of edits!
> 
> Would you mind adding to the documentation of `NOXSolver::solve_with_jacobian` that recovery is possible with the right flag set? Also, you might want to put a changelog entry into `doc/news/changes/minor`! Let me know once you're done with this and then I can merge!

Thank you for the review. Sure, I will do that right after Christmas.

**Comment by vovannikov** [comment]
Done! @bangerth, could you please have a look?

**Comment by bangerth** [comment]
Great, thank you! Last request: Would you mind either squashing the ten commits into one, or into one for the code/documentation changes/one for the new tests/one for the changelog entry?

**Comment by vovannikov** [comment]
> Great, thank you! Last request: Would you mind either squashing the ten commits into one, or into one for the code/documentation changes/one for the new tests/one for the changelog entry?

@bangerth, sure, done.

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Excellent, thank you for your contribution and in particular patience with the process!

**Pull request merged**